Call eglPresentationTimeANDROID when available.

tpsiaki opened the pull request.

tpsiaki commented:
It's been my experience that it works best for the app to be able to set the frame time so that at the other end of MediaCodec, frames can still be identified with what was rendered in them.  For example, in an AR app, one might want the camera's capture time used as the timestamp, which wouldn't necessarily correlate with the wall clock time in the renderer.  Of course, doing it that way would be significantly more plumbing.

pixelflinger commented on filament/src/driver/opengl/OpenGLDriver.cpp:
```diff
@@ -2488,6 +2488,7 @@ void OpenGLDriver::beginFrame(uint64_t monotonic_clock_ns, uint32_t frameId) {
             }
         }
     }
+    mContextManager.setPresentationTime(monotonic_clock_ns);
```
I don't know if this clock is correct. What's passed here is the steady_clock and I don't think this one has anything to do with the system time. What clock does this API expect?

Wouldn't it be easier to NOT add a new method and instead just do it where we do eglSwapBuffers() and use the system time?

romainguy commented:
`ContextManagerCocoa` and `ContextManagerWebGL` are missing `setPresentationTime()`.

pixelflinger commented:
I don't think we need the extra flexibility at this point. I'm not
confident this API is what we want. So I prefer to do something automatic
that doesn't add new apis for now.

On Thu, Sep 27, 2018, 20:21 Romain Guy <notifications@github.com> wrote:

> I believe the spec requires the system monotonic clock. I thought about
> doing this in swap buffers too but this actually gives more flexibility
> (for instance to schedule the frames at a chosen pace/framerate).
>
> The change breaks on Mac and WebGL though.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/filament/pull/315#issuecomment-425308823>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABLvQL4PtPE94jtUaZB2gevII8iW73mSks5ufZWfgaJpZM4W9vIJ>
> .
>

tpsiaki commented:
Added WebGL & Cocoa stubs.

I also added EGL_RECORDABLE_ANDROID to the config which is the last part needed to make this work with MediaCodec surfaces.

pixelflinger commented on filament/src/driver/opengl/ContextManagerEGL.cpp:
```diff
@@ -537,6 +537,7 @@ std::unique_ptr<Driver> ContextManagerEGL::createDriver(void* const sharedGLCont
             EGL_GREEN_SIZE,  8,
             EGL_BLUE_SIZE,   8,
             EGL_ALPHA_SIZE,  0, // reserved to set ALPHA_SIZE below
+            EGL_RECORDABLE_ANDROID, 1,
```
That's okay, but if we can't find a config with RECORDABLE_ANDROID, then we need to try again without it (and maybe print a warning in the log).

tpsiaki replied:
done.

pixelflinger commented on filament/src/driver/opengl/ContextManagerEGL.cpp:
```diff
@@ -63,6 +63,9 @@ using namespace utils;
 namespace filament {
 using namespace driver;
 
+using PFNEGLGETNATIVECLIENTBUFFERANDROIDPROC =
```
Why was this added?

tpsiaki commented:
PFNEGLGETNATIVECLIENTBUFFERANDROIDPROC was giving me an error when building for android locally, but presumably it's a setup thing since it works on kokoro without.  Removed.
